crates/ibc/src/core/ics02_client/handler/update_client.rs

```diff
@@ -19,14 +21,117 @@ use crate::timestamp::Timestamp;
 /// The result following the successful processing of a `MsgUpdateAnyClient` message. Preferably
 /// this data type should be used with a qualified name `update_client::Result` to avoid ambiguity.
 #[derive(Clone, Debug, PartialEq)]
-pub struct Result {
+pub struct UpdateClientResult {
     pub client_id: ClientId,
     pub client_state: Box<dyn ClientState>,
     pub consensus_state: Box<dyn ConsensusState>,
     pub processed_time: Timestamp,
     pub processed_height: Height,
 }
 
+pub(crate) fn validate<Ctx>(ctx: &Ctx, msg: MsgUpdateClient) -> Result<(), Error>
+where
+    Ctx: ValidationContext,
+{
+    let MsgUpdateClient {
+        client_id,
+        header,
+        signer: _,
+    } = msg;
+
+    // Read client type from the host chain store. The client should already exist.
+    // Read client state from the host chain store.
+    let client_state = ctx.client_state(&client_id)?;
+
+    if client_state.is_frozen() {
+        return Err(Error::client_frozen(client_id));
+    }
+
+    // Read consensus state from the host chain store.
+    let latest_consensus_state = ctx
+        .consensus_state(&client_id, client_state.latest_height())
+        .map_err(|_| {
+            Error::consensus_state_not_found(client_id.clone(), client_state.latest_height())
+        })?;
+
+    debug!("latest consensus state: {:?}", latest_consensus_state);
+
+    let now = ctx.host_timestamp();
+    let duration = now
+        .duration_since(&latest_consensus_state.timestamp())
+        .ok_or_else(|| {
+            Error::invalid_consensus_state_timestamp(latest_consensus_state.timestamp(), now)
+        })?;
+
+    if client_state.expired(duration) {
+        return Err(Error::header_not_within_trust_period(
+            latest_consensus_state.timestamp(),
+            now,
+        ));
+    }
+
+    // Use client_state to validate the new header against the latest consensus_state.
+    // This function will return the new client_state (its latest_height changed) and a
+    // consensus_state obtained from header. These will be later persisted by the keeper.
+    let _ = client_state
+        .check_header_and_update_state(ctx, client_id.clone(), header)
+        .map_err(|e| Error::header_verification_failure(e.to_string()))?;
+
+    Ok(())
+}
+
+pub(crate) fn execute<Ctx>(ctx: &mut Ctx, msg: MsgUpdateClient) -> Result<(), Error>
+where
+    Ctx: ExecutionContext,
+{
+    let MsgUpdateClient {
+        client_id,
+        header,
+        signer: _,
+    } = msg;
+
+    // Read client type from the host chain store. The client should already exist.
+    // Read client state from the host chain store.
+    let client_state = ctx.client_state(&client_id)?;
+
+    let UpdatedState {
+        client_state,
+        consensus_state,
+    } = client_state
+        .check_header_and_update_state(ctx, client_id.clone(), header.clone())
```

Comment:
I think we shouldn't be running the complete header verification here and should only decode the header to obtain new client and consensus states. https://github.com/anoma/namada/blob/main/shared/src/ledger/ibc/handler.rs#L1187

Reply:
Right. Should we do the `check_header_and_update_state()` split from #173 first then? Or perhaps merge this but open an issue, which we address before the next release?

Reply:
I am okay with both, but I would prefer the second option (i.e. merge this and open an issue).